### ISSUE TITLE
AlpinoCorpusReader delivers words out of sequence

### ISSUE THREAD

**Issue by LenRemmerswaal** [nltk/nltk#1031]
I noticed that `nltk.corpus.alpino.tagged_sents()` delivers sentences with the words out of order.
Investigation learned that, due to a lexical tree structure existing in the alpino xml element tree, the word elements can appear out of word order in the corpus file, while the alpino reader assumes that the order of word elements coincides with the sentence's word order.
So I fixed that using the patch below, coming out of the local Hg repository that I placed over the nltk folder as installed by a regular recent WinPython installation.  
I did not do any methodical testing. I just used it for training a brill tagger, and noticed the oddities that are now gone. Use this as you see fit. You may want to have `_word()` do its own `SORTTAGWORD.findall()` and sorting to avoid an extra pass over everything.

Versions:
- nltk 3.0.2
- python 2.7.9

The patch:

```
diff -r dcebb5dac345 corpus/reader/bracket_parse.py
--- a/corpus/reader/bracket_parse.py    Tue Jun 30 10:10:43 2015 -0400
+++ b/corpus/reader/bracket_parse.py    Thu Jul 02 16:43:14 2015 -0400
@@ -18,6 +18,7 @@
 from nltk.corpus.reader.api import *

 # we use [^\s()]+ instead of \S+? to avoid matching ()
+SORTTAGWRD = re.compile(r'\((\d+) ([^\s()]+) ([^\s()]+)\)') 
 TAGWORD = re.compile(r'\(([^\s()]+) ([^\s()]+)\)')
 WORD = re.compile(r'\([^\s()]+ ([^\s()]+)\)')
 EMPTY_BRACKETS = re.compile(r'\s*\(\s*\(')
@@ -97,9 +98,9 @@
             return Tree('S', self._tag(t))

     def _tag(self, t, tagset=None):
-        tagged_sent = [(w,t) for (t,w) in TAGWORD.findall(self._normalize(t))]
+        tagged_sent = [(w,p) for (p,w) in TAGWORD.findall(self._normalize(t))]
         if tagset and tagset != self._tagset:
-            tagged_sent = [(w, map_tag(self._tagset, tagset, t)) for (w,t) in tagged_sent]
+            tagged_sent = [(w, map_tag(self._tagset, tagset, p)) for (w,p) in tagged_sent]
         return tagged_sent

     def _word(self, t):
@@ -165,6 +166,12 @@
 class AlpinoCorpusReader(BracketParseCorpusReader):
     """
     Reader for the Alpino Dutch Treebank.
+    This corpus has a lexical breakdown structure embedded, as read by _parse
+    Unfortunately this puts punctuation and some other words out of the sentence
+    order in the xml element tree. This is no good for tag_ and word_
+    _tag and _word will be overridden to use a non-default new parameter 'ordered'
+    to the overridden _normalize function. The _parse function can then remain 
+    untouched.
     """
     def __init__(self, root, encoding='ISO-8859-1', tagset=None):
         BracketParseCorpusReader.__init__(self, root, 'alpino\.xml',
@@ -172,13 +179,47 @@
                                  encoding=encoding,
                                  tagset=tagset)

-    def _normalize(self, t):
+    def _normalize(self, t, ordered = False):
+        """Normalize the xml sentence element in t.
+        The sentence elements <alpino_ds>, although embedded in a few overall 
+        xml elements, are seperated by blank lines. That's how the reader can 
+        deliver them one at a time.
+        Each sentence has a few category subnodes that are of no use to us.
+        The remaining word nodes may or may not appear in the proper order.
+        Each word node has attributes, among which:
+        - begin : the position of the word in the sentence
+        - pos   : Part of Speech: the Tag
+        - word  : the actual word
+        The return value is a string with all xml elementes replaced by 
+        clauses: either a cat clause with nested clauses, or a word clause.
+        The order of the bracket clauses closely follows the xml.
+        If ordered == True, the word clauses include an order sequence number.
+        If ordered == False, the word clauses only have pos and word parts.
+        """
         if t[:10] != "<alpino_ds":
             return ""
         # convert XML to sexpr notation
         t = re.sub(r'  <node .*? cat="(\w+)".*>', r"(\1", t)
-        t = re.sub(r'  <node .*? pos="(\w+)".*? word="([^"]+)".*/>', r"(\1 \2)", t)
+        if ordered:
+            t = re.sub(r'  <node. *?begin="(\d+)".*? pos="(\w+)".*? word="([^"]+)".*?/>', r"(\1 \2 \3)", t)
+        else: 
+            t = re.sub(r'  <node .*?pos="(\w+)".*? word="([^"]+)".*?/>', r"(\1 \2)", t)
         t = re.sub(r"  </node>", r")", t)
         t = re.sub(r"<sentence>.*</sentence>", r"", t)
         t = re.sub(r"</?alpino_ds.*>", r"", t)
         return t
+
+    def _tag(self, t, tagset=None):
+        tagged_sent = [(int(o), w, p) for (o,p,w) in SORTTAGWRD.findall(self._normalize(t, ordered = True))]
+        tagged_sent.sort()
+        if tagset and tagset != self._tagset:
+            tagged_sent = [(w, map_tag(self._tagset, tagset, p)) for (o,w,p) in tagged_sent]
+        else:
+            tagged_sent = [(w,p) for (o,w,p) in tagged_sent]
+        return tagged_sent
+
+    def _word(self, t):
+        """Return a correctly ordered list if words"""
+        tagged_sent = self._tag(t)
+        return [w for (w,p) in tagged_sent]      
+
```

**Comment by LenRemmerswaal** [comment]
Patch now in pull request #1037

**Comment by alexisdimi** [comment]
Nice to see this fixed! The Alpino corpus includes the sentence in the original order (as plain text), in the span <sentence>. You could use it to test the code on the entire corpus for correctness.

**Comment by stevenbird** [comment]
Closed by #1037.
